library/pkcs7_write.c

```diff
@@ -0,0 +1,829 @@
+#include <stdlib.h>
+#define mbedtls_free free
+#define mbedtls_calloc calloc
+#define mbedtls_printf printf
+#define mbedtls_snprintf snprintf
+#endif
+
+/* STRUCTURE OF PKCS7 AND CORRESPONDING FUNCTIONS THAT HANDLE THEM:
+ *PKCS7 {
+ *	CONSTRUCTED | SEQUENCE 										->pkcs7_set_pkcs7_oid
+ *		OID (Signed Data) 										->^
+ *		CONSTRUCTED | CONTEXT SPECIFIC 							->pkcs7_set_signed_data
+ *			CONSTRUCTED | SEQUENCE 								->pkcs7_set_version
+ *				INTEGER (version) 								->^
+ *				CONSTRUCTED |SET 								->pkcs7_set_algo_id
+ *					OID (hash Alg)								->^
+ *				CONSTRUCTED | SEQUENCE 							->pkcs7_set_signed_data_oid
+ *					OID (PKCS7 Data)							->^
+ *				CONSTRUCTED | CONTEXT SPECIFIC 					->pkcs7_set_signer_certs
+ *					entire certificate for each signer  		->^
+ *				CONSTRUCTED | SET 								->pkcs7_set_signers_data
+ *					CONSTRUCTED | SEQUENCE (for each signer) 	->pkcs7_set_each_signer_data
+ *						INTEGER (signedInfoVersion) 			->pkcs7_set_signer_cert_data
+ *						CONSTRUCTED | SEQUENCE 					->^
+ *							certificate issuer info 			->^
+ *							certificate serial					->^
+ *						CONSTRUCTED | SEQUENCE 					->pkcs7_set_algorithm_ids
+ *							OID (hash Alg)						->^
+ *						CONSTRUCTED | SEQUENCE 					->^
+ *							OID (Signature Alg (RSA)) 			->^
+ *						OCTET STRING (signature) 				->pkcs7_set_signature
+ * }
+ */
+
+//struct for containing information needed to generate the pkcs7
+//used for less function arguments
+typedef struct mbedtls_pkcs7_info {
+    unsigned char **crts;  // signing crt DER
+    size_t *crt_sizes;
+    unsigned char **keys;  // signing key DER
+    size_t *key_sizes;
+    int key_pairs;
+    const unsigned char *data;
+    int data_size;
+    mbedtls_md_type_t hash_funct;
+    const char *hash_funct_oid;
+
+} mbedtls_pkcs7_info;
+
+/*
+ *Taken from MBEDTLS, mbedtls-mbedtls2.23.0/programs/util/pem2der.c many thanks
+ *to these great folks Some things were changed though, like memory allocation
+ *THIS ALLOCATES MEMORY, FREE SOMETIME AFTER CALLING
+ */
+int mbedtls_convert_pem_to_der(const unsigned char *input, size_t ilen,
+                       unsigned char **output, size_t *olen)
+{
+    int ret;
+    const unsigned char *s1, *s2, *end = input + ilen;
+    size_t len = 0;
+
+    s1 = (unsigned char *) strstr( (const char *) input, "-----BEGIN" );
+    if( s1 == NULL ) return (-1);
+
+    s2 = (unsigned char *) strstr( (const char *) input, "-----END" );
+    if( s2 == NULL ) return (-1);
+
+    s1 += 10;
+    while( s1 < end && *s1 != '-') 
+        s1++;
+    while( s1 < end && *s1 == '-') 
+        s1++;
+    if( *s1 == '\r') 
+        s1++;
+    if( *s1 == '\n') 
+        s1++;
+
+    if( s2 <= s1 || s2 > end) 
+    	return ( -1 );
+
+    ret = mbedtls_base64_decode( NULL, 0, &len, (const unsigned char *) s1, s2 - s1);
+    if( ret == MBEDTLS_ERR_BASE64_INVALID_CHARACTER ) 
+    {
+        return ( ret );
+    }
+    // free this ouside of function
+    *output = malloc( len );
```

Comment:
Please consistently use only `mbedtls_calloc()` and `mbedtls_free()` - this allows people to replace those with functions of their choice.